src/build/patches.ts

```diff
@@ -0,0 +1,75 @@
+import { parse } from "kdljs";
+import { Enum } from "./types";
+import { readdir, readFile } from "fs/promises";
+import { merge } from "./helpers.js";
+
+interface EnumDescriptor {
+  [x: string]: Enum;
+}
+
+/**
+ * Converts patch files in KDL to match the [types](types.d.ts).
+ */
+export function parseKDL(kdlText: string) {
+  const { output, errors } = parse(kdlText);
+
+  if (errors.length) {
+    throw new Error("KDL parse errors", { cause: errors });
+  }
+
+  const nodes = output!;
+  const enums: EnumDescriptor = {};
+
+  for (const node of nodes) {
+    if (node.name === "enum") {
+      // Handle enum
+      const enumName = node.values[0];
+      if (typeof enumName !== "string") {
+        throw new Error("Missing enum name");
+      }
+      const name = enumName;
+      const values: string[] = [];
+
+      for (const child of node.children ?? []) {
+        if (child.name !== "value" || typeof child.values[0] !== "string") {
+          throw new Error(
+            "enum values should be in the form of `value {name}`",
+          );
+        }
+        values.push(child.values[0]);
+      }
+
+      enums[enumName] = { name, value: values };
+    }
+  }
+
+  return { enums: { enum: enums } };
+}
+
+/**
+ * Recursively collect all KDL file URLs in a directory.
+ */
+async function getAllKDLFileURLs(folder: URL): Promise<URL[]> {
+  const entries = await readdir(folder, { withFileTypes: true });
+  return entries.map((entry) => new URL(entry.name, folder));
+}
+
+/**
+ * Read and parse a single KDL file.
+ */
+export async function readInputKDL(fileUrl: URL): Promise<any> {
```

Comment:
And this can be readPatch. (You probably followed the existing readInputJSON, which is literally just parsing JSON and returned the result as-is, but in this case we also convert the data so it doesn't fit much.)